### PR TITLE
feat: Elasticsearch vector database

### PR DESCRIPTION
**Pull request by HaoXuAI**
- Add ElasticSearch in the online store as another vector database.
- Adjust the `retrieve_online_document` API
# What this PR does / why we need it:

# Which issue(s) this PR fixes:

# Fixes

### CODE REVIEW

File: sdk/python/feast/infra/online_stores/contrib/elastichsearch.py [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+
+from feast import Entity, FeatureView, RepoConfig
+from feast.infra.key_encoding_utils import get_list_val_str, serialize_entity_key
+from feast.infra.online_stores.online_store import OnlineStore
+from feast.protos.feast.types.EntityKey_pb2 import EntityKey as EntityKeyProto
+from feast.protos.feast.types.Value_pb2 import Value as ValueProto
+from feast.repo_config import FeastConfigBaseModel
+
+
+class ElasticsearchOnlineStoreConfig(FeastConfigBaseModel):
```
Comment (franciscojavierarceo) [inline sdk/python/feast/infra/online_stores/contrib/elastichsearch.py:17]: nice

File: sdk/python/feast/feature_store.py [hunk 2]
```diff
@@ -1886,7 +1886,7 @@ def retrieve_online_documents(
         feature: str,
         query: Union[str, List[float]],
         top_k: int,
-        distance_metric: str,
+        distance_metric: Optional[str] = None,
```
Comment (HaoXuAI) [inline sdk/python/feast/feature_store.py:1889]: have to make it optional as `elasticsearch` doesn't allow specifying the metric in the online API. Instead, it has to update the index.

**Review by franciscojavierarceo — APPROVED** [review]
Nice 🚀🚀🚀

File: sdk/python/feast/feature_store.py (hunk 2 above)
Comment (franciscojavierarceo) [reply sdk/python/feast/feature_store.py:1889]: Makes sense can we add some explicit tests for this to highlight the behavior? Also this is awesome.

**Pull request merged**